### PR TITLE
CI: Add per-dependency build scripts and update CI workflow

### PR DESCRIPTION
**Pull request by PatTheMav**
### Description

This PR adds universal binary support for all pre-compiled OBS dependencies. This is achieved via the following changes:

* Each dependency's build step has been moved into its separate build script
* Build scripts use a pre-defined template for patching/building/preparing to keep necessary customisations per-dependency low
* Each dependency can be built standalone (via its own script) and allows explicit specification of the build architecture (`x86_64`, `arm64`, or `universal`), the host's architecture is used by default
* Dylib dependency have their absolute paths replaced by `@rpath` references, pre-existing `LC_RPATH` values are removed before packaging or creation of universal binaries via `lipo`.
* Separate scripts to generate universal binaries given `arm64` and `x86_64` build trees
* Patches for dependencies are now kept as patch files within the repository

There are still warts with this PR (mainly the macOS version baked into some libraries is not correct) and the workflow itself needs to be tested on CI of course.

### Motivation and Context
This change allows OBS to be built and packaged as either a universal or platform-specific build and also limits the housekeeping for each dependency to its own build script, reducing the maintenance and update effort.

### How Has This Been Tested?
* Builds for all three variants ((`x86_64`, `arm64`, and `universal`) built locally with fresh as well as existing temporary build directory 
* OBS built with x86 libraries
* OBS built with universal libraries (detected as Apple M1 binary on an M1 MacBook Air)

### Types of changes
- New feature (non-breaking change which adds functionality)
- Code cleanup (non-breaking change which makes code smaller or more readable)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by PatTheMav** [comment]
As always CI breakage with issues that never occurred on local systems, will have to investigate why Qt ignores the c++17 switch and x264 all of a sudden doesn't want to cross-compile arm64 anymore.

@Developer-Ecosystem-Engineering feel free to take a look.

File: utils/build_script_generator.py
Comment (PatTheMav) [inline utils/build_script_generator.py:40]: @derrod lmk if you know of an easier pattern to get get pairs of `A_VERSION` and `A_HASH` into a single list containing only `A` while also preserving order.. 😬

Comment (derrod) [reply utils/build_script_generator.py:40]: Wouldn't a good old for-loop be simpler?

Something like
```python
dependency_strings = []

for key in filtered_data.keys():
    prefix = key.partition('_')[0]
    dependency_strings.append(
        f'    "{prefix.lower()}'
        f' {environment_data.get(f"{prefix}_VERSION", "")}'
        f' {environment_data.get(f"{prefix}_HASH", "")}"'
    )
```

A normal python list preserves order anyway so I'm not entirely sure why you're essentially creating a list from the `dict_keys` view, only to then turn that into a dict again, which then gets used in a `dict_keys` view again. The list from `[key.split("_")[0] for key in filtered_data.keys()]` could be used directly.

Comment (derrod) [reply utils/build_script_generator.py:40]: Or wait, the issue is that things aren't unique, right?

Comment (PatTheMav) [reply utils/build_script_generator.py:40]: The double conversion eliminates duplicates, that's the trick. By creating a dict from the keys minus the suffix, I eliminate the duplicates, but still preserve order. `set(dict.keys()` shuffles the order, so that was out of the question.. 😂

Comment (derrod) [reply utils/build_script_generator.py:40]: In that case the only thing I really see is just using a generator instead of creating a temporary list, and combining the two into one big mess. A for-loop over a dictionary will also iterate over keys by default so you don't need to explicitly do `.keys()`

```python
dependency_strings = [
    (
        f'    "{key.lower()}'
        f" {environment_data.get(f'{key}_VERSION', '')}"
        f" {environment_data.get(f'{key}_HASH', '')}\""
    )
    for key in dict.fromkeys(key.partition('_')[0] for key in filtered_data)
]
```

**Comment by PatTheMav** [comment]
@RytoEX @DDRBoxman - if we could get this reviewed, merged and released first, I could adapt pending PRs to these new deps and add preparations for universal/multi-arch builds. I got a branch prepared for that, but that's obviously based on my rework PR for now.

**Comment by RytoEX** [comment]
> 
> 
> @RytoEX @DDRBoxman - if we could get this reviewed, merged and released first, I could adapt pending PRs to these new deps and add preparations for universal/multi-arch builds. I got a branch prepared for that, but that's obviously based on my rework PR for now.

I'd looked over the PR quickly when it was submitted and it seemed fine at a glance, though I haven't gone over it in depth.  I'm much in favor of this approach over the YAML -> script approach.  I think my one request at the moment would be that if patches are OS-agnostic, that they be located in a global patches directory instead of sequestered in OS-specific sub-directories.

**Comment by PatTheMav** [comment]
> I think my one request at the moment would be that if patches are OS-agnostic, that they be located in a global patches directory instead of sequestered in OS-specific sub-directories.

I agree, but as we have no other OSes or even "global" patches right now we should keep this in mind when we add Windows deps building.

File: .github/workflows/create_scripts.yml
Comment (RytoEX) [inline .github/workflows/create_scripts.yml:6]: There is no workflow with this name.  Is that intentional?  What does the `create_scripts.yml` do now, anyway?

**Comment by RytoEX** [comment]
> 
> 
> > I think my one request at the moment would be that if patches are OS-agnostic, that they be located in a global patches directory instead of sequestered in OS-specific sub-directories.
> 
> I agree, but as we have no other OSes or even "global" patches right now we should keep this in mind when we add Windows deps building.

The QTBUG-74606.patch affects Windows.  We build with it currently.

File: .github/workflows/create_scripts.yml
Comment (PatTheMav) [reply .github/workflows/create_scripts.yml:6]: Right, it should say `main` - will push a fix in a bit.

The script now reads the dependency config in the workflow and updates the global build script, so we don't have to update the versions/hashes in two files.

The global build script is convenience for local builds, so I thought this would make maintenance convenient for us as well as keeping that script up2date whenever we do a release.

Comment (RytoEX) [reply .github/workflows/create_scripts.yml:6]: Ah, so now it reads https://github.com/obsproject/obs-deps/pull/60/files#diff-7829468e86c1cc5d5133195b5cb48e1ff6c75e3e9203777f6b2e379d9e4882b3R23-R56 and updates https://github.com/obsproject/obs-deps/pull/60/files#diff-c1f2aefe3a1708caf92d65020c7b456155cb3c8b8f7dbc69a8f0cb1076130893R25-R42 ?  Is that about it?  Does that necessitate having the exact same deps for each OS we build for?

Comment (PatTheMav) [reply .github/workflows/create_scripts.yml:6]: Not in that specific case, because the script uses the values set as `env` in the OS-specific build job. So if you specify different versions for Windows, it wouldn't change a thing.

It might trip up the standalone scripts' fallbacks (as they are read from the workflow and take the first value they encounter), but I'll pass that bridge once we get there. I'll probably have to use more complex parsing in Powershell anyway so that becomes a non-issue.

Comment (RytoEX) [reply .github/workflows/create_scripts.yml:6]: Sounds good for now.

**Comment by PatTheMav** [comment]
> > > I think my one request at the moment would be that if patches are OS-agnostic, that they be located in a global patches directory instead of sequestered in OS-specific sub-directories.
> > 
> > 
> > I agree, but as we have no other OSes or even "global" patches right now we should keep this in mind when we add Windows deps building.
> 
> The QTBUG-74606.patch affects Windows. We build with it currently.

Got it, will relocate it together with the fix for the script creation workflow then.

File: .github/workflows/create_scripts.yml
Comment (PatTheMav) [reply .github/workflows/create_scripts.yml]: Should be fixed with a recent force-push.

File: README.md
Comment (RytoEX) [inline README.md:11]: This link is not for xiph.org, but rather their GitHub mirror.

File: .github/workflows/main.yml
Comment (RytoEX) [inline .github/workflows/main.yml:326]: Would this make more sense as its own script instead of raw script in yaml?

**Review by RytoEX — CHANGES_REQUESTED** [review]
Just a quick visual review.  Nothing really stood out other than some stuff in the README.  Someone else will have to handle actually doing test runs on the build scripts.  While it'd be good to test these locally, ultimately we'll also have see how it works after merging and have someone test the resulting dependencies.  Unfortunately, I can do neither of these steps.

I'm not sure if I like having everything in a "CI" directory, but I can't think of an alternative that is better, so I'm not going to fuss about it.  A "scripts" directory doesn't sound much better, and if it's going to contain patches, it's also a bit confusing.  Conceptually, I'd prefer if the patches were in a root-level subdirectory, but maybe in practice that doesn't work as well.  I'll leave that to you.

File: .github/workflows/main.yml
Comment (PatTheMav) [reply .github/workflows/main.yml:326]: It is implemented in the packaging scripts but those contain much more convenience functionality that wasn't needed here.

File: .github/workflows/main.yml
Comment (RytoEX) [inline .github/workflows/main.yml:27]: I understand why it's appealing to set the version to 2021-05-12.  However, if you run the `update_version` script included in the xiph/opus repo with this commit checked out, it will output a `package_version` file with "1.3.1-93-gdfd6c88a" as its set package version.  It makes more sense to use the authoritative version info rather than making up our own.  We should either leave this at 1.3.1, or use the version info generated by their script.

File: .github/workflows/main.yml
Comment (RytoEX) [inline .github/workflows/main.yml:57]: Should this be sorted alphabetically so there's some order to it?

File: README.md
Comment (RytoEX) [inline README.md:10]: See comment above about specific opus version.

File: README.md
Comment (RytoEX) [inline README.md:68]: Nitpick:  GitHub

This improper capitalization exists elsewhere in this PR as well (dependency version table), but in places that already exist.  Those other instances can easily be corrected separately.

File: README.md
Comment (RytoEX) [inline README.md:71]: `OS'` doesn't need an apostrophe by itself.  If it were `OS's`, then maybe, but not required, according to APA guidelines.  I'll leave the choice between `OS` vs. `OSs` vs. `OS's` up to you.

**Review by RytoEX — CHANGES_REQUESTED** [review]
Mostly nitpicks.

I'd like to see some additional confirmation that the deps produced from this PR work fine for building OBS on macOS.  This PR is generating artifacts, so anyone with a Mac who can build OBS should be able to help with this.  Keeping track of what version of macOS and what host arch was used would be helpful.

**Comment by Developer-Ecosystem-Engineering** [comment]
> As always CI breakage with issues that never occurred on local systems, will have to investigate why Qt ignores the c++17 switch and x264 all of a sudden doesn't want to cross-compile arm64 anymore.
> 
> @Developer-Ecosystem-Engineering feel free to take a look.

Was able to build + run against obs-studio/master using the arm64 dependencies generated with this change 

- MacBook Air M1
- macOS 11.2.3 
- Clang 12.0.5

Was able to capture from built in camera, record using the default .mkv and .mp4. 

obs-apple-silicon

File: .github/workflows/main.yml
Comment (PatTheMav) [reply .github/workflows/main.yml:57]: Sorted by build order which is also in parts dependant on quirks in how FFmpeg dependencies need to be built (i.e. MbedTLS and SRT).

**Review by OsvaldoRino — APPROVED** [review]

File: CI/build-qt-macos.sh
Comment (DDRBoxman) [inline CI/build-qt-macos.sh:7]: Incorrect description

**Review by DDRBoxman — CHANGES_REQUESTED** [review]

**Pull request merged**
